hail/python/hail/methods/statgen.py

```diff
@@ -1697,6 +1699,323 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
+    n = A.take(1)[0].ndarray.shape[1]
+
+    # Generate random matrix G
+    G = np.random.normal(0, 1, (n,l))
+
+    # Helper Functions
+
+    def chunk_ndarray(Q, local_part_sizes):
+        groups = []
+        for part in local_part_sizes:
+            start = part.rows_preceeding
+            end = start + part.part_size
+            Q_i = Q[start:end]
+            groups.append(Q_i)
+        return groups
+
+    def concatToNumpy(A):
+        blocks = A.ndarray.collect()
+        big_mat = np.concatenate(blocks, axis=0)
+        return big_mat
+
+    def ndarray_to_table(chunked_arr):
+        structs = [hl.struct(row_group_number = hl.int64(idx), ndarray = block)
+                   for idx, block in enumerate(chunked_arr)]
+        ht = hl.Table.parallelize(structs)
+        ht = ht.key_by('row_group_number')
+        return ht
+
+    def hailBlanczos(A, G, k, l, q, block_size, times):
+
+        start = time.time()
+
+        h_list = []
+        G_i = G
+
+        for j in range(0, q):
+            temp = A.annotate(H_i = A.ndarray @ G_i)
+            temp = temp.annotate(G_i_intermediate = temp.ndarray.T @ temp.H_i)
+            result = temp.aggregate(hl.struct(Hi_chunks = hl.agg.collect(temp.H_i), 
+                                                G_i = hl.agg.ndarray_sum(temp.G_i_intermediate)))
+            localized_H_i = np.vstack(result.Hi_chunks)
+            h_list.append(localized_H_i)
+            G_i = result.G_i
+
+        temp = A.annotate(H_i = A.ndarray @ G_i)
+        result = temp.aggregate(hl.agg.collect(temp.H_i))
+        localized_H_i = np.vstack(result)
+        h_list.append(localized_H_i)
+
+        H = np.hstack(h_list) 
+
+        end = time.time()
+        q_loop_time = end - start
+
+        start = time.time()
+        
+        # perform QR decomposition on unblocked version of H
+        Q, R = np.linalg.qr(H)
+
+        end = time.time()
+        process_Q_time = end - start
+
+        A = A.annotate(part_size = A.ndarray.shape[0])
+        A = A.annotate(rows_preceeding = hl.int32(hl.scan.sum(A.part_size)))
+        A = A.annotate_globals(Qt = Q.T)
+        T = A.annotate(ndarray = A.Qt[:, A.rows_preceeding:A.rows_preceeding + A.part_size] @ A.ndarray)
+        arr_T = T.aggregate(hl.agg.ndarray_sum(T.ndarray))
+        
+        start = time.time()
+
+        U, S, W = np.linalg.svd(arr_T, full_matrices=False)
+
+        end = time.time()
+        svd_time = end - start
+
+        start = time.time()
+
+        V = Q @ U
+
+        end = time.time()
+
+        get_V_time = end - start
+        
+        truncV = V[:,:k]
+        truncS = S[:k]
+        truncW = W[:k,:]
+            
+        if times:
+            return truncV, truncS, truncW, q_loop_time, process_Q_time, svd_time, get_V_time
+        else:
+            return truncV, truncS, truncW
+
+    U, S, V = hailBlanczos(A, G, k, l, q, block_size, report_times)
+    matrix_S = np.diag(S)
+
+    scores = V.transpose() @ matrix_S
```

Comment:
Get rid of `matrix_S`. No reason to manifest the matrix. Faster to do `scores = V.transpose() * S`, gives same result.